File: src/main/java/me/hhitt/disasters/disaster/impl/RedLightGreenLight.java
package me.hhitt.disasters.disaster.impl;

import me.hhitt.disasters.Disasters;
import me.hhitt.disasters.arena.Arena;
import me.hhitt.disasters.disaster.Disaster;
import me.hhitt.disasters.service.DeathMessageService;
import me.hhitt.disasters.util.Msg;
import me.hhitt.disasters.util.Notify;
import org.bukkit.Bukkit;
import org.bukkit.event.EventHandler;
import org.bukkit.event.HandlerList;
import org.bukkit.event.Listener;
import org.bukkit.event.player.PlayerMoveEvent;

import java.util.ArrayList;
import java.util.List;

public class RedLightGreenLight implements Disaster, Listener {

    private final List<Arena> arenas = new ArrayList<>();
    private boolean phaseRed = false;
    private int lastPhaseChange = 0;

    @Override
    public void start(Arena arena) {
        arenas.add(arena);
        phaseRed = false;
        lastPhaseChange = 0;
        Bukkit.getPluginManager().registerEvents(this, Disasters.getInstance());
        Notify.disaster(arena, "red-light-green-light");
    }

    @Override
    public void pulse(int time) {
        if (time - lastPhaseChange >= 4) {
            phaseRed = !phaseRed;
            lastPhaseChange = time;
            String title = phaseRed ? "<red><bold>RED LIGHT" : "<green><bold>GREEN LIGHT";
            for (Arena arena : new ArrayList<>(arenas)) {
                for (org.bukkit.entity.Player p : arena.getAlive()) {
                    Msg.sendTitle(p, title);
                }
            }
        }
    }

    @Override
    public void stop(Arena arena) {
        arenas.remove(arena);
        HandlerList.unregisterAll(this);
    }

    @EventHandler
    public void onMove(PlayerMoveEvent event) {
        if (!phaseRed) return;
        Arena arena = null;
        for (Arena a : arenas) {
            if (a.isPlayerValid(event.getPlayer())) {
                arena = a;
                break;
            }
        }
        if (arena == null) return;
        double dx = event.getTo().getX() - event.getFrom().getX();
        double dy = event.getTo().getY() - event.getFrom().getY();
        double dz = event.getTo().getZ() - event.getFrom().getZ();
        if (dx * dx + dy * dy + dz * dz > 0.0009) {
            event.setCancelled(true);
            DeathMessageService.mark(event.getPlayer(), "red-light-green-light");
            event.getPlayer().getWorld().strikeLightningEffect(event.getTo());
            event.getPlayer().damage(20.0);
        }
    }
}

File: src/main/kotlin/me/hhitt/disasters/disaster/impl/RedLightGreenLight.kt
package me.hhitt.disasters.disaster.impl

import me.hhitt.disasters.arena.Arena
import me.hhitt.disasters.disaster.Disaster
import me.hhitt.disasters.service.DeathMessageService
import me.hhitt.disasters.util.Msg
import me.hhitt.disasters.util.Notify
import org.bukkit.Bukkit
import org.bukkit.Sound
import org.bukkit.event.EventHandler
import org.bukkit.event.HandlerList
import org.bukkit.event.Listener
import org.bukkit.event.player.PlayerMoveEvent

class RedLightGreenLight : Disaster, Listener {

    private val arenas = mutableListOf<Arena>()
    private var phaseRed = false
    private var lastPhaseChange = 0

    override fun start(arena: Arena) {
        arenas.add(arena)
        phaseRed = false
        lastPhaseChange = 0
        Bukkit.getPluginManager().registerEvents(this, me.hhitt.disasters.Disasters.getInstance())
        Notify.disaster(arena, "red-light-green-light")
    }

    override fun pulse(time: Int) {
        if (time - lastPhaseChange >= 4) {
            phaseRed = !phaseRed
            lastPhaseChange = time
            val title = if (phaseRed) "<red><bold>RED LIGHT" else "<green><bold>GREEN LIGHT"
            arenas.toList().forEach { arena ->
                arena.alive.forEach { p -> Msg.sendTitle(p, title) }
            }
        }
    }

    override fun stop(arena: Arena) {
        arenas.remove(arena)
        HandlerList.unregisterAll(this)
    }

    @EventHandler
    fun onMove(event: PlayerMoveEvent) {
        if (!phaseRed) return
        val arena = arenas.firstOrNull { it.isPlayerValid(event.player) } ?: return
        val dx = event.to.x - event.from.x
        val dy = event.to.y - event.from.y
        val dz = event.to.z - event.from.z
        if (dx * dx + dy * dy + dz * dz > 0.0009) {
            event.isCancelled = true
            DeathMessageService.mark(event.player, "red-light-green-light")
            event.player.world.strikeLightningEffect(event.to)
            event.player.damage(20.0)
        }
    }
}
